pallene/reaching_definitions.lua

```diff
@@ -0,0 +1,377 @@
+
+local ir = require "pallene.ir"
+local location = require "pallene.location"
+
+--local inspect = require "inspect"
+--local ppi = function(t,f) return inspect(t,{newline='',process=f}) end
+--local pp = function(t,f) return inspect(t,{process=f}) end
+
+local RD = {}
+-- start set operations -----------
+
+local function copy(t1,t2)
+    for k,v in pairs(t2) do t1[k] = v end
+end
+local function deduplicate(t)
+    local hash = {}
+    local res = {}
+    for _,v in ipairs(t) do
+       if not hash[v] then
+           res[#res+1] = v
+           hash[v] = true
+       end
+    end
+    return res
+end
+local function union(t1, t2)
+    local ret = {}
+    for _,v in ipairs(t1) do
+        table.insert(ret,v)
+    end
+    for _,v in ipairs(t2) do
+        table.insert(ret,v)
+    end
+    return deduplicate(ret)
+end
+local function difference(t1,t2)
+    local ret = {}
+    local hash = {}
+    for _,v in ipairs(t2) do
+       if not hash[v] then
+           hash[v] = true
+       end
+    end
+    for _,v in ipairs(t1) do
+        if not hash[v] then
+            table.insert(ret,v)
+        end
+    end
+    return ret
+end
+local function equal(t1,t2)
+    if #t1 ~= #t2 then return false end
+    local h1,h2 = {},{}
+    for k,v in ipairs(t1) do
+       if not h1[v] then
+           h1[v] = true
+       end
+       if not h2[t2[k]] then
+           h2[t2[k]] = true
+       end
+    end
+    for k,_ in pairs(h1) do
+        if not h2[k] then return false end
+    end
+    return true
+end
+
+
+-- end   set operations -----------
+
+-- start control flow -------------
+local function newnode(cmd,cfg,parent)
+    table.insert(cfg,{
+        cmd = cmd,
+        to={},
+        from=parent
+    })
+    for _,par in ipairs(parent) do
+        if par > 0 then
+            table.insert(cfg[par].to,#cfg)
+        end
+    end
+    return {#cfg}
+end
+local function build_cfg(cmd,cfg,parent)
+    local tag = cmd._tag
+    if     tag == "ir.Cmd.Nop" then
+        --nothing
+    elseif tag == "ir.Cmd.Seq" then
+        local newidx = parent
+        for _, c in ipairs(cmd.cmds) do
+            newidx = build_cfg(c,cfg,newidx)
+        end
+        return newidx
+    elseif tag == "ir.Cmd.If" then
+        local newidx = build_cfg(cmd.condition,cfg,parent)
+        local then_ = cmd.then_ and build_cfg(cmd.then_,cfg,newidx) or {}
+        local else_ = cmd.else_ and build_cfg(cmd.else_,cfg,newidx) or {}
+        return union(then_,else_)
+    elseif tag == "ir.Cmd.Loop" then
+        return build_cfg(cmd.body,cfg,parent)
+    elseif tag == "ir.Cmd.For" then
+        local newidx = parent
+        newidx = build_cfg(cmd.start,cfg,newidx)
+        newidx = build_cfg(cmd.limit,cfg,newidx)
+        newidx = build_cfg(cmd.step,cfg,newidx)
+        newidx = build_cfg(cmd.body,cfg,newidx)
+        return newidx
+    elseif string.match(tag, "^ir%.Cmd%.") then
+        return newnode(cmd,cfg,parent)
+    elseif string.match(tag, "^ir%.Value%.") then
```

Comment:
it doesn't generate a node for Values, it simply returns the `parent` variable so that the function can continue. Initially I though of putting an `else` there, but I wanted to explicitly say to not create a node on the case of values

Reply:
What is happening when you call build_cfg on cmd.start and cmd.limit, in the case for ir.Cmd.For? Isn't that building cfg nodes for values?

> but I wanted to explicitly say not to create a node for values.

The `error("impossible")` that we put after the else already serves that purpose.

If you want a clearer error message during development, I'd suggest to put that inside the else instead of as another elseif case. That makes it clear that it is a run-time type error.

Reply:
> What is happening when you call build_cfg on cmd.start and cmd.limit, in the case for ir.Cmd.For? Isn't that building cfg nodes for values?

I thought that they were expressions. But after you explain me that in this point there isn't expressions (only values), I changed it.